test/jdk/sun/security/rsa/TestKeyFactory.java

```diff
@@ -63,55 +141,79 @@ private static void testKey(Key key1, Key key2) throws Exception {
             }
         } else if (key1 instanceof PrivateKey) {
             if (key2.getFormat().equals("PKCS#8") == false) {
-                throw new Exception("Format not PKCS#8");
+                throw new Exception("Format not PKCS#8: " + key2.getFormat());
             }
         }
-        if (key1.equals(key2) == false) {
-            throw new Exception("Keys not equal");
+        // skip equals check when key1 is custom key
+        if (key1 != CUSTOM_PRIV && key1 != CUSTOM_PUB) {
+            if (!key1.equals(key2)) {
+                throw new Exception("Keys not equal");
+            }
         }
-        if (Arrays.equals(key1.getEncoded(), key2.getEncoded()) == false) {
-            throw new Exception("Encodings not equal");
+
+        // only compare encodings if keys are of the same format
+        if (key1.getFormat().equals(key2.getFormat())) {
+            if (!Arrays.equals(key1.getEncoded(), key2.getEncoded())) {
+                throw new Exception("Encodings not equal");
+            }
         }
```

Comment:
Can we combine the 2 blocks above into one? That is to say, when key1 and key2 have the same format, we check the equality of both getEncoded() and themselves. Same for the P11 test.

Reply:
Sure, will do.